language-plutus-core/docs/Builtins.md

```diff
@@ -0,0 +1,340 @@
+charBoolInteger =
+    Proxy @Char `TypeSchemeArrow`
+    Proxy @Bool `TypeSchemeArrow`
+    TypeSchemeResult (Proxy @Integer)
+```
+
+`FoldArgs` takes the list of types of arguments of a function and the type of the result and folds them back into the type of that function. E.g. `FoldArgs [Char, Bool] Integer` gives `Char -> Bool -> Integer` back.
+
+Finally, `BuiltinFunctionMeaning` is just a `TypeScheme` of a primitive function and the primitive function itself (of type `FoldArgs args res`).
+
+Then when we encounter `AppBuiltinFunction fun args` during evaluation we evaluate `fun` and look up its meaning in the provided environment of `BuiltinFunctionMeaning`s, which allows to retrieve a `TypeScheme`, as well as a Haskell function `f`. Having the `TypeScheme` we know what type each of `args` must have, so after evaluating them to constants we traverse that list of constants and check if the type tag of every constant is the same one as is currently expected from the `TypeScheme` info: if the tags match, then the types they represent must also match and the Haskell type checker sees that ('cause we're using GADTs), so we can type-safely apply `f` to a freshly obtained constant.
+
+The `TypeScheme` of a built-in function also allows to get the Plutus Core type of that function: it's straightforward recursion on `TypeScheme` where each `TypeSchemeArrow` becomes `->` of the target language and each implicit ``uni `Includes` arg`` representing an argument becomes a `BuiltinType (SomeType uni)`, and ``uni `Includes` res`` representing the resulting type also becomes a `BuiltinType (SomeType uni)`.
+
+So having a single `TypeScheme` is enough to adapt the type checker and the evaluator to extensible built-in functions.
+
+## Saturated vs unsaturated built-in functions
+
+There are two ways to have built-in function in the AST of `Term`:
+
+1. either it's saturated application `AppBuiltinFunction BuiltinFunction [Term]` (i.e. the one that we've been using so far) where a built-in function is always applied to the exact number of arguments that it expects, e.g. `AddInteger` has to always be applied to two arguments
+
+2. or it's unsaturated application:
+
+```haskell
+data Term
+    = <...>
+    | BuiltinFunction BuiltinFunction
+    | Apply Term Term
+```
+
+where a built-in function can be partially applied, e.g. it's allowed to apply `AddInteger` to just one argument and pass the partially applied function as an argument to some higher-order function like `map`
+
+Those two representations are very similar. Differences:
+
+- with unsaturated built-in functions we can write `map (addInteger 1) xs` while with saturated once we have to write `map (\x -> addInteger 1 x) xs`
+- it seems that it's easier to implement an efficient evaluator with saturated built-in fuctions
+
+In general, differences are minor, so either of the approaches is fine. Currently we have (1), but we're going to change it to (2) at some point.
+
```

Comment:
Wait!  I thought we were going with saturated builtins:  see #1943 and https://jira.iohk.io/browse/SCP-250.  Isn't that (1)?

Reply:
Yes, we currently have (2) and I think we want (1).

Reply:
Also apropos partial application, something that always worries me is how you tell the difference between `f x y = x+y` and `g x = \y -> x+y`, since they both have the same type (that's definitely problematic in the impure world, where you have the opportunity to insert an effect between x and y in the second version) .  If we had partial application of builtins,  would that present any difficulties wrt compilation? I guess the current mechanism handles that OK though.  Maybe if we had saturated application and higher-order builtins then that would require extra care, since if you're compiling from a higher-level language you'd have to check where the "real" arguments of (something that maps to) a builtin finish and ones that belong to the returned function begin (and that could happen arbitrarily many times).

Reply:
> That's (1) isn't it, so don't we want to change from (2) to (1)?

Yeah, I just can't count.